<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Requirements &mdash; IMPACT-Pipeline 1.0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="IMPACT-Pipeline 1.0.1 documentation" href="index.html" />
    <link rel="next" title="Usage" href="usage.html" />
    <link rel="prev" title="Installation" href="install.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="usage.html" title="Usage"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="install.html" title="Installation"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">IMPACT-Pipeline 1.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="requirements">
<h1>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h1>
<div class="sidebar">
<p class="first sidebar-title">Note:</p>
<p class="last">This will only run on any Linux system that has an SGE or LSF cluster.</p>
</div>
<p>Please see template.conf file in the configuration folder.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">perl:</th><td class="field-body"><a class="reference external" href="http://perl5.git.perl.org/perl.git/tag/2c93aff028f866699beb26e5e7504e531c31b284">v5.20.2</a></td>
</tr>
<tr class="field-even field"><th class="field-name">python:</th><td class="field-body"><a class="reference external" href="https://www.python.org/download/releases/2.7.8/">v2.7.8</a></td>
</tr>
<tr class="field-odd field"><th class="field-name">R:</th><td class="field-body"><a class="reference external" href="http://cran.r-project.org/src/base/R-3/R-3.1.2.tar.gz">v3.1.2</a></td>
</tr>
</tbody>
</table>
<div class="section" id="required-perl-packages">
<h2>Required Perl Packages<a class="headerlink" href="#required-perl-packages" title="Permalink to this headline">¶</a></h2>
<div class="sidebar">
<p class="first sidebar-title">Note:</p>
<p class="last">For <strong>MSKCC_DMP_Logger</strong> is wriiten internally. To get this please contact us.</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Getopt::Long:</th><td class="field-body"><a class="reference external" href="http://perldoc.perl.org/Getopt/Long.html">Getopt::Long</a></td>
</tr>
<tr class="field-even field"><th class="field-name">IO::File:</th><td class="field-body"><a class="reference external" href="http://perldoc.perl.org/IO/File.html">IO::File</a></td>
</tr>
<tr class="field-odd field"><th class="field-name">Tie::Autotie:</th><td class="field-body"><a class="reference external" href="http://search.cpan.org/~pinyan/Tie-Autotie-0.03/lib/Tie/Autotie.pm">Tie::Autotie &#8216;Tie::IxHash&#8217;</a></td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">Statistics::Lite:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body"><a class="reference external" href="http://search.cpan.org/~brianl/Statistics-Lite-3.62/Lite.pm">Statistics::Lite qw(:all)</a></td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">MSKCC_DMP_Logger:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body"><a class="reference external" href="mailto:rons&#46;shah&#37;&#52;&#48;gmail&#46;com">Contact Us</a></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="required-python-packages">
<h2>Required Python Packages<a class="headerlink" href="#required-python-packages" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">pandas:</th><td class="field-body"><a class="reference external" href="http://pandas.pydata.org/">Pandas</a></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="required-r-packages">
<h2>Required R Packages<a class="headerlink" href="#required-r-packages" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">knitr:</th><td class="field-body"><a class="reference external" href="http://yihui.name/knitr/">knitr</a></td>
</tr>
<tr class="field-even field"><th class="field-name">ggplot2:</th><td class="field-body"><a class="reference external" href="http://ggplot2.org/">ggplot2</a></td>
</tr>
<tr class="field-odd field"><th class="field-name">plyr:</th><td class="field-body"><a class="reference external" href="https://cran.r-project.org/web/packages/plyr/index.html">pylr</a></td>
</tr>
<tr class="field-even field"><th class="field-name">reshape:</th><td class="field-body"><a class="reference external" href="https://cran.r-project.org/web/packages/reshape/index.html">reshape</a></td>
</tr>
<tr class="field-odd field"><th class="field-name">scales:</th><td class="field-body"><a class="reference external" href="https://cran.r-project.org/web/packages/scales/index.html">scales</a></td>
</tr>
<tr class="field-even field"><th class="field-name">xtable:</th><td class="field-body"><a class="reference external" href="https://cran.r-project.org/web/packages/xtable/index.html">xtable</a></td>
</tr>
<tr class="field-odd field"><th class="field-name">RColorBrewer:</th><td class="field-body"><a class="reference external" href="https://cran.r-project.org/web/packages/RColorBrewer/index.html">RColorBrewer</a></td>
</tr>
<tr class="field-even field"><th class="field-name">gplots:</th><td class="field-body"><a class="reference external" href="https://cran.r-project.org/web/packages/gplots/index.html">gplots</a></td>
</tr>
<tr class="field-odd field"><th class="field-name">googleVis:</th><td class="field-body"><a class="reference external" href="https://cran.r-project.org/web/packages/googleVis/index.html">googleVis</a></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="purpose-and-tools-used">
<h2>Purpose and Tools Used<a class="headerlink" href="#purpose-and-tools-used" title="Permalink to this headline">¶</a></h2>
<div class="sidebar">
<p class="first sidebar-title">Note:</p>
<p class="last">For Timmomatic we are using a custom old version which uses <a class="reference external" href="https://cutadapt.readthedocs.org">cutadapt v1.1</a> internally. To get this please contact us.</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Trimming:</th><td class="field-body"><a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic</a></td>
</tr>
<tr class="field-even field"><th class="field-name">Alignment:</th><td class="field-body"><a class="reference external" href="https://github.com/lh3/bwa/tree/0.7.5a">BWA v0.7.5a</a></td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">Somatic SNV calling:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body"><a class="reference external" href="https://github.com/broadinstitute/mutect/tree/1.1.4">MuTect v1.1.4</a></td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">Somatic INDEL calling:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body"><a class="reference external" href="http://www.broadinstitute.org/gatk/download">SomaticIndelDetector in GATK v2.3-9</a></td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">Somatic INDEL calling:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body"><a class="reference external" href="https://github.com/genome/pindel/tree/v0.2.5a7">PINDEL v0.2.5a7</a></td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">Indel Realignment:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body"><a class="reference external" href="https://github.com/mozack/abra/tree/v0.92">ABRA v0.92</a></td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">Mark Duplicates and Various Statistics:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body"><a class="reference external" href="https://github.com/broadinstitute/picard/tree/1.96">Picard Tools v1.96</a></td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">Base Quality Recalibration and Find Covered Intervals:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body"><a class="reference external" href="http://www.broadinstitute.org/gatk/download">GATK v3.3.0</a></td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">Genotyping Position:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body"><a class="reference external" href="https://github.com/samtools/samtools/tree/0.1.19">SAMTOOLS v0.1.19</a></td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">Somatic Structural Variant Framework:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body"><a class="reference external" href="https://github.com/rhshah/IMPACT-SV/tree/1.0.1">IMPACT-SV v1.0.1</a></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="inside-the-config-file">
<h2>Inside the config file<a class="headerlink" href="#inside-the-config-file" title="Permalink to this headline">¶</a></h2>
<p>There are three sections:</p>
<table border="1" class="docutils">
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Section 1</th>
<th class="head">Section 2</th>
<th class="head">Section 3</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Locations</td>
<td>Parameters</td>
<td>Versions</td>
</tr>
</tbody>
</table>
<p>All of this section start with <code class="docutils literal"><span class="pre">&gt;</span></code> sign.</p>
<p>Inside each of the section here are the things that need to be set:</p>
<div class="section" id="locations">
<h3>Locations<a class="headerlink" href="#locations" title="Permalink to this headline">¶</a></h3>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">ZCAT:</th><td class="field-body">Location of the <code class="docutils literal"><span class="pre">zcat</span></code> program on linux</td>
</tr>
<tr class="field-even field"><th class="field-name">TMPDIR:</th><td class="field-body">Set the temporary directory for all tools please set somthing other then <code class="docutils literal"><span class="pre">/tmp</span></code></td>
</tr>
<tr class="field-odd field"><th class="field-name">JAVA_1_6:</th><td class="field-body">Set JAVA version 1.6</td>
</tr>
<tr class="field-even field"><th class="field-name">JAVA_1_7:</th><td class="field-body">Set JAVA version 1.7</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">GATK_SomaticIndel:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Path to GATK somatic indel detector (GATK version 2.3-9)</td>
</tr>
<tr class="field-even field"><th class="field-name">GATK:</th><td class="field-body">Path to GATK (GATK version 3.3.0)</td>
</tr>
<tr class="field-odd field"><th class="field-name">Reference:</th><td class="field-body">Path to fasta referece file to be used (GRCh37)</td>
</tr>
<tr class="field-even field"><th class="field-name">Refseq:</th><td class="field-body">Path to refgene file to be used</td>
</tr>
<tr class="field-odd field"><th class="field-name">PICARD:</th><td class="field-body">Path to picard tools (Picard version 1.19)</td>
</tr>
<tr class="field-even field"><th class="field-name">Mutect:</th><td class="field-body">Path to MuTect (MuTect version 1.1.4)</td>
</tr>
<tr class="field-odd field"><th class="field-name">BaitInterval:</th><td class="field-body">Bait file to be used for analysis. Please make the interval file based on Picard&#8217;s HSmetrics tool format.</td>
</tr>
<tr class="field-even field"><th class="field-name">TargetInterval:</th><td class="field-body">Target file to be used for analysis. Please make the interval file based on Picard&#8217;s HSmetrics tool format.</td>
</tr>
<tr class="field-odd field"><th class="field-name">ABRA:</th><td class="field-body">Path to ABRA (ABRA version 0.92)</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">TargetRegionLIST:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Target File in GATK&#8217;s List format.</td>
</tr>
<tr class="field-odd field"><th class="field-name">PINDELBIN:</th><td class="field-body">Path to pindel installation (Pindel version 0.2.5a7)</td>
</tr>
<tr class="field-even field"><th class="field-name">SVpipeline:</th><td class="field-body">Path to structural variant framework (only required if running the structural variant detection framework)</td>
</tr>
<tr class="field-odd field"><th class="field-name">CAT:</th><td class="field-body">Location of <code class="docutils literal"><span class="pre">cat</span></code> program on Linux</td>
</tr>
<tr class="field-even field"><th class="field-name">PYTHON:</th><td class="field-body">Location of <code class="docutils literal"><span class="pre">python</span></code> program on Linux (Python version 2.7.8)</td>
</tr>
<tr class="field-odd field"><th class="field-name">TrimGalore:</th><td class="field-body">Path to trimgalore tool</td>
</tr>
<tr class="field-even field"><th class="field-name">PERL:</th><td class="field-body">Location of <code class="docutils literal"><span class="pre">perl</span></code> program on Linux (Perl version 5.20.2)</td>
</tr>
<tr class="field-odd field"><th class="field-name">BWA:</th><td class="field-body">Path to bwa tool</td>
</tr>
<tr class="field-even field"><th class="field-name">GeneInterval:</th><td class="field-body">Gene interval file</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">GeneIntervalAnn:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Gene interval annotated file</td>
</tr>
<tr class="field-even field"><th class="field-name">GeneCoord:</th><td class="field-body">Path to Gene Coordinate file</td>
</tr>
<tr class="field-odd field"><th class="field-name">TilingInterval:</th><td class="field-body">= Path to tiling intervals</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">TilingIntervalAnn:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Path to tiling intervals - annotated for cytoband, for copy number</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">FingerPrintInterval:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Path to FingerPrint Interval file</td>
</tr>
<tr class="field-even field"><th class="field-name">dbSNP:</th><td class="field-body">Path to db snp vcf file</td>
</tr>
<tr class="field-odd field"><th class="field-name">COSMIC:</th><td class="field-body">Path to cosmic vcf (version 0.68)</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">Mills_1000G_Indels:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Path to Mills 1000G Indels</td>
</tr>
<tr class="field-odd field"><th class="field-name">dbSNP_bitset:</th><td class="field-body">Path to dbsnp bitset file</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">AnnotateAssessFilterVariants:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Path to Annotate Assess and Filter variants script</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">LoessNormalization:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Path ot Loess Normalization for copynumber</td>
</tr>
<tr class="field-even field"><th class="field-name">GCBiasFile:</th><td class="field-body">Path to GCbias file for copy number</td>
</tr>
<tr class="field-odd field"><th class="field-name">HistNormDir:</th><td class="field-body">Path to Historiacal Normal dir for Copy number</td>
</tr>
<tr class="field-even field"><th class="field-name">BestCopyNumber:</th><td class="field-body">Path to Copy number script</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">NormVsNormCopyNumber:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Path to Normal vs. Normal Copy number script</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">StdNormalLoess_TM:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Standard Normals for copy number analysis - FFPE for tumor samples#</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">StdNormalLoess_NVN:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Standard Normals for copy number normal vs normal analysis</td>
</tr>
<tr class="field-even field"><th class="field-name">AllMetrics:</th><td class="field-body">Path to all metrics R script</td>
</tr>
<tr class="field-odd field"><th class="field-name">SAMTOOLS:</th><td class="field-body">Path to samtools</td>
</tr>
<tr class="field-even field"><th class="field-name">BEDTOOLS:</th><td class="field-body">Path to bedtools</td>
</tr>
<tr class="field-odd field"><th class="field-name">GenotypeAllele:</th><td class="field-body">Path to Genotype allele script</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">CosmicHotspotVcf:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Path to cosmic hotspot vcf</td>
</tr>
<tr class="field-odd field"><th class="field-name">Annovar:</th><td class="field-body">Path to Annovar script</td>
</tr>
<tr class="field-even field"><th class="field-name">Annovar_db:</th><td class="field-body">Path to Annovar DB</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">Canonical_refFlat_file:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Path to canonical reflat file</td>
</tr>
<tr class="field-even field"><th class="field-name">IGVtools:</th><td class="field-body">Path to IGV tools</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">TranslationFolder:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Path to translation folder</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">HotSpot_mutations:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Path to hotspot mutations for 2 tiered filtering</td>
</tr>
<tr class="field-odd field"><th class="field-name">clinicalExons:</th><td class="field-body">ListOfClinicalExon</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">Validated_Exons:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">File with List Of Clinically Validated Exons</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">Tumor_supressor_list:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Path to list of tumor supressor genes</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">Canonical_Exon_Interval_table_with_aa:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Path to exon interval table</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">Canonical_Exon_Interval_list:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Path to canonical exon interval table for DoC</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">NormalVariantsVCF:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Path to compiled variants found in mixed normals</td>
</tr>
<tr class="field-odd field"><th class="field-name">QSUB:</th><td class="field-body">Path to qsub for SGE</td>
</tr>
<tr class="field-even field"><th class="field-name">BSUB:</th><td class="field-body">Path to bsub for LSF</td>
</tr>
<tr class="field-odd field"><th class="field-name">RHOME:</th><td class="field-body">Path to R bin directory</td>
</tr>
<tr class="field-even field"><th class="field-name">RLIBS:</th><td class="field-body">Path to R library directory</td>
</tr>
<tr class="field-odd field"><th class="field-name">RSYNC:</th><td class="field-body">Path to <code class="docutils literal"><span class="pre">rsyn</span></code> on system</td>
</tr>
<tr class="field-even field"><th class="field-name">BarcodeKey:</th><td class="field-body">Path to barcode key file</td>
</tr>
<tr class="field-odd field"><th class="field-name">AdaptorKey:</th><td class="field-body">Path to adaptor key file</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">StandardNormalsDirectory:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Directory where the standard normals are stored</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="parameters">
<h3>Parameters<a class="headerlink" href="#parameters" title="Permalink to this headline">¶</a></h3>
<p>Set the parameters to different file/folders/values required by the IMPACT pipeline</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name" colspan="2">StdNormalForMutationCalling:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Path to standard normal to be used for mutation calling</td>
</tr>
<tr class="field-even field"><th class="field-name">ListOfFiles:</th><td class="field-body">File of Files(FOF) for different steps for the pipeline (only required when the process dont start from merging fastq)</td>
</tr>
<tr class="field-odd field"><th class="field-name">Process:</th><td class="field-body">Which process to run the pipeline on ( can be 1,2,3,4,5,6,7 independently or continuous combination in ascending order )</td>
</tr>
<tr class="field-even field"><th class="field-name">FastqSource:</th><td class="field-body">Where are the fastq file from (can be <code class="docutils literal"><span class="pre">GCL</span></code> or <code class="docutils literal"><span class="pre">DMP</span></code>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">MAPQ:</th><td class="field-body">Mapping Quality Threshold (Used by DMP-IMPACT:0.2)0</td>
</tr>
<tr class="field-even field"><th class="field-name">BASQ:</th><td class="field-body">Base Quality Threshold (Used by DMP-IMPACT:0.2)</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">MergeDinucleotide:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Flag to Merge di-nucleotide mutation(can be 1(True) or 2(False))</td>
</tr>
<tr class="field-even field"><th class="field-name">MoveFiles:</th><td class="field-body">Flag to Move file in folders (can be 1(True) or 2(False))</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">DeleteIntermediateFiles:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Flag ti Delete Intermediate Files (can be 1(True) or 2(False))</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">TNfreqRatio_MutectStdFilter:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">TN freq Ratio for mutect std filter (Used by DMP-IMPACT:5)</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">TNfreqRatio_SomIndelStdFilter:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">TN freq Ratio for SID std filter (Used by DMP-IMPACT:5)</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">VF_threshold_hotspot:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Variant Frequency threshold for SNV hotspot (Used by DMP-IMPACT:0.01)</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">AD_SomIndelSTDFilter:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Allele Depth Threshold for SID standard filter (Used by DMP-IMPACT:5)</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">DP_SomIndelSTDFilter:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Total Depth Threshold for SID standard filter (Used by DMP-IMPACT:0)</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">VF_SomIndelSTDilter:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Variant Frequency Threshold for SID standard filter (Used by DMP-IMPACT:0.01)</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">AD_MutectSTDFilter:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Allele Depth Threshold for Mutect standard filter (Used by DMP-IMPACT:5)</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">DP_MutectSTDFilter:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Total Depth Threshold for Mutect standard filter (Used by DMP-IMPACT:0)</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">VF_MutectSTDFilter:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Variant Frequency Threshold for Mutect standard filter (Used by DMP-IMPACT:0.01)</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">TNfreqRatio_AnnotationFilter:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Tumor to Normal frequency ratio therehold for Annotation (Used by DMP-IMPACT:5)</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">PON_AD_Threshold:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Panel of Normal Allele Depth Threshold (Used by DMP-IMPACT:3)</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">PON_TPVF_Threshold:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Panel of Normal TPVF Threshold (Used by DMP-IMPACT:10)</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">Pindel_Min_Indel_Len:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Minimum Length of INDEL called by PINDEL(Used by DMP-IMPACT:25)</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">Pindel_Max_Indel_Len:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Maximum Length of INDEL called by PINDEL (Used by DMP-IMPACT:2000)</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">MAFthreshold_AnnotationFilter:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Maf threshold for Annotation (Used by DMP-IMPACT:0.01)</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">minimumDPforSNV:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Minimum Total Depth for Novel SNVs  (Used by DMP-IMPACT:20)</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">minimumADforSNV:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Minimum Allele Depth for Novel SNVs (Used by DMP-IMPACT:10)</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">minimumVFforSNV:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Minimum Variant Frequency for Novel SNVs (Used by DMP-IMPACT:0.05)</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">minimumDPforSNVhs:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Minimum Total Depth for Hotspot SNVs (Used by DMP-IMPACT:20)</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">minimumADforSNVhs:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Minimum Allele Depth for Hotspot SNVs (Used by DMP-IMPACT:8)</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">minimumVFforSNVhs:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Minimum Variant Frequency for Hotspot SNVs (Used by DMP-IMPACT:0.02)</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">minimumDPforINDEL:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Minimum Total Depth for Novel INDELs (Used by DMP-IMPACT:20)</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">minimumADforINDEL:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Minimum Allele Depth for Novel INDELs (Used by DMP-IMPACT:10)</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">minimumVFforINDEL:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Minimum Variant Frequency for Novel INDELs (Used by DMP-IMPACT:0.05)</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">minimumDPforINDELhs:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Minimum Total Depth for Hotspot INDELs (Used by DMP-IMPACT:20)</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">minimumADforINDELhs:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Minimum Allele Depth for Hotspot INDELs (Used by DMP-IMPACT:8)</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">minimumVFforINDELhs:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Minimum Variant Frequnecy for Hotspot INDELs (Used by DMP-IMPACT:0.02)</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">occurrencePercent:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Minimum Percentage For Occurrence In Other Normals (Used by DMP-IMPACT:0.2)</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">Coverage_threshold_darwin_report:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">Coverage threshold for darwin reports(good coverage vs bad coverage) (Used by DMP-IMPACT:100)</td>
</tr>
<tr class="field-odd field"><th class="field-name">QUEUE_NAME:</th><td class="field-body">Name of the queue on the SGE or LSF</td>
</tr>
<tr class="field-even field"><th class="field-name">CLUSTER:</th><td class="field-body">Flag for what cluster to be used (can <code class="docutils literal"><span class="pre">SGE</span></code> or <code class="docutils literal"><span class="pre">LSF</span></code>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">runABRA:</th><td class="field-body">Flag to whether use ABRA or GATK indel realignment(can be 1(True) or 2(False))</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="versions">
<h3>Versions<a class="headerlink" href="#versions" title="Permalink to this headline">¶</a></h3>
<div class="sidebar">
<p class="first sidebar-title">Note:</p>
<p class="last">This section is just to print what version of things you are using so you can have all the dependencies with the respective versions listed here.</p>
</div>
<p>Inside the version there are version that are being used for each tool. This is just for consistency in reports.</p>
</div>
</div>
<div class="section" id="description-for-title-file-txt">
<h2>Description for title_file.txt<a class="headerlink" href="#description-for-title-file-txt" title="Permalink to this headline">¶</a></h2>
<p>Headers for this tab-delimited file should be exactly with this names:</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Barcode:</th><td class="field-body">Has to start with bc and end with any number [for example: bc01 or bc101 should match the <strong>adaptor &amp; barcode</strong> file mentioned in configuration file</td>
</tr>
<tr class="field-even field"><th class="field-name">Pool:</th><td class="field-body">Can be any string <strong>joined by ``-``</strong> and <strong>not ``_``</strong> and all entries should be from same pool</td>
</tr>
<tr class="field-odd field"><th class="field-name">Sample_ID:</th><td class="field-body">Can be any string <strong>joined by ``-``</strong> and <strong>not ``_``</strong></td>
</tr>
<tr class="field-even field"><th class="field-name">Collab_ID:</th><td class="field-body">Can be any string or <code class="docutils literal"><span class="pre">-</span></code></td>
</tr>
</tbody>
</table>
<div class="sidebar">
<p class="first sidebar-title">Note:</p>
<p class="last">Patient with multiple samples should have <strong>same Patient_ID</strong></p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Patient_ID:</th><td class="field-body">Can be any string <strong>joined by ``-``</strong> and <strong>not ``_``</strong></td>
</tr>
<tr class="field-even field"><th class="field-name">Class:</th><td class="field-body">Can be Tumor or Normal.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Sample_type:</th><td class="field-body">Can be any string or <code class="docutils literal"><span class="pre">-</span></code></td>
</tr>
<tr class="field-even field"><th class="field-name">Input_ng:</th><td class="field-body">Can be any float or <code class="docutils literal"><span class="pre">-</span></code></td>
</tr>
<tr class="field-odd field"><th class="field-name">Library_yield:</th><td class="field-body">Can be float or <code class="docutils literal"><span class="pre">-</span></code></td>
</tr>
<tr class="field-even field"><th class="field-name">Pool_input:</th><td class="field-body">Can be float or <code class="docutils literal"><span class="pre">-</span></code></td>
</tr>
<tr class="field-odd field"><th class="field-name">Bait_version:</th><td class="field-body">Can be any string or <code class="docutils literal"><span class="pre">-</span></code></td>
</tr>
<tr class="field-even field"><th class="field-name">Gender:</th><td class="field-body">Can be any Male/Female or <code class="docutils literal"><span class="pre">-</span></code></td>
</tr>
<tr class="field-odd field"><th class="field-name">PatientName:</th><td class="field-body">Can be any string or <code class="docutils literal"><span class="pre">-</span></code></td>
</tr>
<tr class="field-even field"><th class="field-name">MAccession:</th><td class="field-body">Can be any string or <code class="docutils literal"><span class="pre">-</span></code></td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">Extracted_DNA_Yield:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">Can be a float or <code class="docutils literal"><span class="pre">-</span></code></td>
</tr>
</tbody>
</table>
<p>For analysis to start the <strong>outputDirectory</strong> will be required to have this file with <code class="docutils literal"><span class="pre">title_file.txt</span></code> as the name or this file needs to be present in the <strong>configuration</strong> file with either <code class="docutils literal"><span class="pre">title_file.txt</span></code> as then name or <code class="docutils literal"><span class="pre">Pool_title.txt</span></code> as the name where <strong>Pool</strong> is the string used above for that category.</p>
</div>
<div class="section" id="description-for-samplesheet-csv">
<h2>Description for SampleSheet.csv<a class="headerlink" href="#description-for-samplesheet-csv" title="Permalink to this headline">¶</a></h2>
<p>This is a comma separated file is created by the illumina sequencer and it is used to merge the fastq files.</p>
<p>Headers for this tab-delimited file should be exactly with this names:</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">FCID:</th><td class="field-body">Flowcell ID (required)</td>
</tr>
<tr class="field-even field"><th class="field-name">Lane:</th><td class="field-body">Lane Number, this is used to merge the fastq files across lanes (required)</td>
</tr>
<tr class="field-odd field"><th class="field-name">SampleID:</th><td class="field-body">Sample ID, this is used to merge the files (required)</td>
</tr>
<tr class="field-even field"><th class="field-name">SampleRef:</th><td class="field-body">Sample Reference is from [example:HUMAN]</td>
</tr>
<tr class="field-odd field"><th class="field-name">Index:</th><td class="field-body">Index used to sequence the sample (require)</td>
</tr>
<tr class="field-even field"><th class="field-name">Description:</th><td class="field-body">Description of the samples</td>
</tr>
<tr class="field-odd field"><th class="field-name">Control:</th><td class="field-body">Can be any string or <code class="docutils literal"><span class="pre">-</span></code></td>
</tr>
<tr class="field-even field"><th class="field-name">Recipe:</th><td class="field-body">Can be any string or <code class="docutils literal"><span class="pre">-</span></code></td>
</tr>
<tr class="field-odd field"><th class="field-name">Operator:</th><td class="field-body">Can be any string or <code class="docutils literal"><span class="pre">-</span></code></td>
</tr>
<tr class="field-even field"><th class="field-name">SampleProject:</th><td class="field-body">Can be any string or <code class="docutils literal"><span class="pre">-</span></code></td>
</tr>
</tbody>
</table>
<p>For analysis to start the <strong>outputDirectory</strong> will be required to have this file with <code class="docutils literal"><span class="pre">SampleSheet.csv</span></code> as the name or this file needs to be present in the <strong>configuration</strong> file with <code class="docutils literal"><span class="pre">SampleSheet.csv</span></code> as the name.</p>
</div>
<div class="section" id="description-for-adaptor-file-in-the-configuration-file">
<h2>Description for adaptor file in the configuration file<a class="headerlink" href="#description-for-adaptor-file-in-the-configuration-file" title="Permalink to this headline">¶</a></h2>
<p>The adaptor file is the tab-delimited file with two columns:</p>
<ol class="arabic simple">
<li>Barcode Key to which the adaptor belongs which should always start with <code class="docutils literal"><span class="pre">bc</span></code></li>
<li>Adaptor sequence itself</li>
</ol>
<p>There is <strong>no header</strong> in this file.</p>
<p>For Example:</p>
<blockquote>
<div><table border="1" class="docutils">
<colgroup>
<col width="9%" />
<col width="91%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>bc01</td>
<td>GATCGGAAGAGCACACGTCTGAACTCCAGTCACAACGTGATATCTCGTATGCCGTCTTCTGCTTG</td>
</tr>
</tbody>
</table>
</div></blockquote>
</div>
<div class="section" id="description-for-barcode-file-in-the-configuration-file">
<h2>Description for barcode file in the configuration file<a class="headerlink" href="#description-for-barcode-file-in-the-configuration-file" title="Permalink to this headline">¶</a></h2>
<p>The barcode file is the tab-delimited file with two columns:</p>
<ol class="arabic simple">
<li>Barcode Sequece</li>
<li>Barcode Number that sequence represent.</li>
</ol>
<p>There is <strong>a header</strong> in this file.</p>
<dl class="docutils">
<dt>For Example:</dt>
<dd><table border="1" class="first last docutils">
<colgroup>
<col width="39%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Sequence</th>
<th class="head">TruSeqBarcode</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>AACGTGAT</td>
<td>bc01</td>
</tr>
</tbody>
</table>
</dd>
</dl>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Requirements</a><ul>
<li><a class="reference internal" href="#required-perl-packages">Required Perl Packages</a></li>
<li><a class="reference internal" href="#required-python-packages">Required Python Packages</a></li>
<li><a class="reference internal" href="#required-r-packages">Required R Packages</a></li>
<li><a class="reference internal" href="#purpose-and-tools-used">Purpose and Tools Used</a></li>
<li><a class="reference internal" href="#inside-the-config-file">Inside the config file</a><ul>
<li><a class="reference internal" href="#locations">Locations</a></li>
<li><a class="reference internal" href="#parameters">Parameters</a></li>
<li><a class="reference internal" href="#versions">Versions</a></li>
</ul>
</li>
<li><a class="reference internal" href="#description-for-title-file-txt">Description for title_file.txt</a></li>
<li><a class="reference internal" href="#description-for-samplesheet-csv">Description for SampleSheet.csv</a></li>
<li><a class="reference internal" href="#description-for-adaptor-file-in-the-configuration-file">Description for adaptor file in the configuration file</a></li>
<li><a class="reference internal" href="#description-for-barcode-file-in-the-configuration-file">Description for barcode file in the configuration file</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="install.html"
                        title="previous chapter">Installation</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="usage.html"
                        title="next chapter">Usage</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/requirements.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="usage.html" title="Usage"
             >next</a> |</li>
        <li class="right" >
          <a href="install.html" title="Installation"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">IMPACT-Pipeline 1.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Ronak H Shah, Donavan Cheng, Ahmet Zehir, Aijazuddin Syed, Raghu Chandramohan, Meera Prasad.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.5.
    </div>
  </body>
</html>